sound/soc/sof/sof-audio.c

```diff
@@ -325,6 +327,63 @@ static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
 		}
 	}
 
+	/*
+	 * The above loop handles connections between widgets that belong to the DAPM widget list.
+	 * But this is not sufficient to handle special cases such as the one needed for setting
+	 * up the route between the feedback pipeline and the playback pipeline for the smart
+	 * amplifier. So handle those separately while checking that both pipelines are in use.
+	 */
+	list_for_each_entry(sroute, &sdev->route_list, list) {
+		bool src_widget_in_dapm_list, sink_widget_in_dapm_list;
+
+		if (sroute->setup)
+			continue;
+
+		src_widget_in_dapm_list = widget_in_list(list, sroute->src_widget->widget);
+		sink_widget_in_dapm_list = widget_in_list(list, sroute->sink_widget->widget);
+
+		/*
+		 * if both source and sink are in the DAPM list, the route must already have been
+		 * set up above. And if neither are in the DAPM list, the route shouldn't be
+		 * handled now.
+		 */
+		if ((src_widget_in_dapm_list && sink_widget_in_dapm_list) ||
+		    (!src_widget_in_dapm_list && !sink_widget_in_dapm_list))
+			continue;
+
+		/*
+		 * Check the use_count of the widget that is not in the DAPM list to confirm if
+		 * it is in use currently before setting up the route.
+		 */
+		if (src_widget_in_dapm_list) {
+			mutex_lock(&sroute->sink_widget->setup_mutex);
+			if (!sroute->sink_widget->use_count) {
+				mutex_unlock(&sroute->sink_widget->setup_mutex);
+				continue;
+			}
+		} else {
+			mutex_lock(&sroute->src_widget->setup_mutex);
+			if (!sroute->src_widget->use_count) {
+				mutex_unlock(&sroute->src_widget->setup_mutex);
+				continue;
+			}
+		}
+
+		if (tplg_ops && tplg_ops->route_setup) {
+			ret = tplg_ops->route_setup(sdev, sroute);
```

Comment:
how would this route be freed? Maybe add a comment that the free part doesn't need special handling because 
````<insert reason here>````